src/engine/guidance/post_processing.cpp

```diff
@@ -292,6 +296,124 @@ std::vector<RouteStep> postProcess(std::vector<RouteStep> steps)
+    // As a direct implication, we have to keep the time of the initial/final turns (which adds a
+    // few seconds of inaccuracy at both ends. This is acceptable, however, since the turn should
+    // usually not be as relevant.
+
+    if (steps.size() <= 2)
+        return;
+
+    // if phantom node is located at the connection of two segments, either one can be selected as
+    // turn
+    //
+    // a --- b
+    //       |
+    //       c
+    //
+    // If a route from b to c is requested, both a--b and b--c could be selected as start segment.
+    // In case of a--b, we end up with an unwanted turn saying turn-right onto b-c.
+    // These cases start off with an initial segment which is of zero length.
+    // We have to be careful though, since routing that starts in a roundabout has a valid
+    // initial segment of length zero and we cannot delete the upcoming segment.
+
+    if (steps.front().distance <= std::numeric_limits<double>::epsilon() &&
+        !entersRoundabout((steps.begin() + 1)->maneuver.instruction))
+    {
+        // We have to adjust the first step both for its name and the bearings
+        const auto &current_depart = steps.front();
+        auto &designated_depart = *(steps.begin() + 1);
+
+        // FIXME this is required to be consistent with the route durations. The initial turn is not
+        // actually part of the route, though
+        designated_depart.duration += current_depart.duration;
+        designated_depart.distance;
+
+        geometry.locations.erase(geometry.locations.begin());
+
+        BOOST_ASSERT(geometry.segment_offsets[1] == 1);
+        // geometry offsets have to be adjusted. Move all offsets to the front and reduce by one.
+        std::transform(geometry.segment_offsets.begin() + 1, geometry.segment_offsets.end(),
+                       geometry.segment_offsets.begin(), [](const std::size_t val)
+                       {
+                           return val - 1;
+                       });
+        geometry.segment_offsets.pop_back();
+
+        // remove the initial distance value
+        geometry.segment_distances.erase(geometry.segment_distances.begin());
+
+        // update initial turn direction/bearings. Due to the duplicated first coordinate, the
+        // initial bearing is invalid
+        designated_depart.maneuver =
+            detail::stepManeuverFromGeometry(TurnInstruction::NO_TURN(), WaypointType::Depart, geometry);
+
+        // finally remove the initial (now duplicated move)
+        steps.erase(steps.begin());
+
+        // and update the leg geometry indices for the removed entry
+        std::for_each(steps.begin(), steps.end(), [](RouteStep &step)
+                      {
+                          --step.geometry_begin;
+                          --step.geometry_end;
+                      });
+    }
+
+    auto &next_to_last_step = *(steps.end() - 2);
```

Comment:
With the erase above only the guard at the top of the function guarantees for this not being an out of bounds access; maybe assert on size here, too just to be sure (e.g. for when we incorporate future changes).